improve backwards compatibility for existing projects

bcressey opened the pull request:
**Issue number:**

Related: #207

**Description of changes:**
The changes queued up amount to flag day where existing projects have to make structural changes and clean existing build artifacts. This is an annoyance for local development, but could be a bigger headache if past point release branches need to move to a newer version of `twoliter` for any reason.

Try to ease some of the pain in the transition by making a few concessions to backwards compatibility:

1. Use existing RPMs during builds if they exist, but clean them up if the corresponding package is rebuilt so the project converges on the new structure over time.
2. Keep the old target directory that cargo uses to track what's been built, but move it to the new location.
3. Support the workspace manifest in either the old location, under `variants`, or the new location at the project root.
4. Treat manifests with no buildsys-related metadata as "packages" (rather than "variants" or "kits").


**Testing done:**
Built the main project repo with the previous release of `twoliter`, updated to a local build of `twoliter` with these changes, and confirmed that builds and cleanup worked as expected. 

Output showing warnings:
```
[cargo-make] INFO - cargo make 0.36.11
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: default
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: install-twoliter
Found Twoliter v0.1.1 installed.
Skipping installation.
[cargo-make] INFO - Execute Command: "/home/fedora/brdev/tools/twoliter/twoliter" "--log-level=info" "make" "default" "--project-path=/home/fedora/brdev/Twoliter.toml" "--cargo-home=/home/fedora/brdev/.cargo" "--"
[2024-05-29T17:22:16Z WARN  twoliter::project] A Release.toml file was found. Release.toml is deprecated. Please remove it from your project.
[cargo-make][1] INFO - Build File: /home/fedora/brdev/build/tools/Makefile.toml
[cargo-make][1] INFO - Task: default
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Running Task: setup
[cargo-make][1] INFO - Running Task: setup-build
[cargo-make][1] INFO - Running Task: fetch-sdk
[cargo-make][1] INFO - Running Task: fetch-sources
[cargo-make][1] INFO - Running Task: fetch-vendored
[cargo-make][1] INFO - Running Task: check-licenses
bans ok, licenses ok, sources ok
[cargo-make][1] INFO - Running Task: fetch-licenses
Skipping fetching licenses
[cargo-make][1] INFO - Running Task: build-sbkeys
[cargo-make][1] INFO - Running Task: publish-setup
17:22:21 [INFO] Found infra config at path: /home/fedora/brdev/Infra.toml
[cargo-make][1] INFO - Running Task: cargo-metadata
Found deprecated project workspace /home/fedora/brdev/variants/Cargo.toml.
Please relocate it to /home/fedora/brdev/Cargo.toml.
[cargo-make][1] INFO - Running Task: build-variant
   Compiling libtirpc v0.1.0 (/home/fedora/brdev/packages/libtirpc)
   Compiling aws-signing-helper v0.1.0 (/home/fedora/brdev/packages/aws-signing-helper)
   Compiling static-pods v0.1.0 (/home/fedora/brdev/packages/static-pods)
   Compiling ecr-credential-provider-1_27 v0.1.0 (/home/fedora/brdev/packages/ecr-credential-provider-1.27)
   Compiling os v0.1.0 (/home/fedora/brdev/packages/os)
warning: netdog@0.1.0: Expected to find one of 'build-package', 'build-kit', or 'build-variant' in package.metadata. Assuming 'build-package'.
warning: netdog@0.1.0: Expected to find one of 'build-package', 'build-kit', or 'build-variant' in package.metadata. Assuming 'build-package'.
   Compiling kmod-6_1-nvidia v0.1.0 (/home/fedora/brdev/packages/kmod-6.1-nvidia)
...
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
buildsys: assume package if missing metadata

Previously, the `build-package` metadata table could have been left
empty if the package did not need any external files. Meanwhile, an
empty `build-variant` for variants would have been useless, as then
no packages would be installed.

Treat manifests with no metadata table as belonging to packages, to
avoid a breaking change.

The `build-kit` metadata table for kits is a new use case which
doesn't create any backwards compatibility risks.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: use older workspace manifest if present

To minimize churn in the boilerplate around a project, accept the old
location for the cargo workspace covering the artifacts that buildsys
manages. Emit a warning in that case to encourage a move to the new
path.

To avoid ambiguity, refuse to build a project that has both files.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: relocate old target directory if present

If the old location for cargo's output directory exists, move it to
the new one to avoid rebuilding the world unnecessarily.

Also, align the target dirs for "build-package" and "build-variant",
so that the records that cargo creates for one task are found if the
other task ever runs.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: accept rpms built with an older version

Newly built packages will be siloed into per-package directories as
part of the new work to enable kits. Older packages may exist in the
parent directory, and can still be used as long as they do not need
to be included in kits.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: expand search for migrations and kernel

When dealing with older build trees, the corresponding packages may
be in the parent directory or in package-specific subdirectories.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: clean package artifacts in old location

We moved the location where package artifacts are stored when adding
support for kits. To avoid breaking cleanup in existing workspaces,
also check the previous location and remove any matching artifacts.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

webern commented on tools/buildsys/src/builder.rs:
```diff
@@ -859,16 +859,18 @@ where
+            )?);
+            output_file.set_extension("");
+            cleanup(&output_file, output_dir, &mut clean_dirs)?;
+            cleanup(&marker_file, build_dir, &mut clean_dirs)?;
```
Is it a problem that this is inside the loop and being called more than once for each marker file since there is now more than one `output_dir/output_file`?

bcressey replied:
> Is it a problem that this is inside the loop and being called more than once for each marker file since there is now more than one output_dir/output_file?

I'll fix that. It's not directly a problem since `cleanup` returns immediately if the file doesn't exist, but it would mean that this sequence breaks:
1. output file doesn't exist in first directory
2. marker file is cleaned up in first iteration
3. output file exists in next directory, but removal fails with an error

The second time the program is run, the marker file would be gone and so the attempt to clean up the output would not be repeated.

webern commented on twoliter/embedded/Makefile.toml:
```diff
@@ -329,11 +329,10 @@ dependencies = ["setup"]
-  cargo fetch --locked --manifest-path "${BUILDSYS_SOURCES_DIR}/Cargo.toml"
-fi
-
-cargo fetch --locked --manifest-path "${BUILDSYS_ROOT_DIR}/Cargo.toml"
```
Edit: (I missed the `.` associated with this commend)

~We still need this or else we are breaking the ability to go to one workspace. Are you saying we will no longer have one workspace as a possibility? It would need to be something like:~

```bash
if [ -f "${BUILDSYS_ROOT_DIR}/Cargo.toml" ]; then
  cargo fetch --locked --manifest-path "${BUILDSYS_ROOT_DIR}/Cargo.toml"
fi
```

webern commented on twoliter/embedded/Makefile.toml:
```diff
@@ -329,11 +329,10 @@ dependencies = ["setup"]
-
-cargo fetch --locked --manifest-path "${BUILDSYS_ROOT_DIR}/Cargo.toml"
+for ws in sources variants .; do
+  [ -s "${BUILDSYS_ROOT_DIR}/${ws}/Cargo.toml" ] || continue
```
Why is this `-s` and not `-f`?

bcressey replied:
I usually prefer `-s` over `-f` for cases where a non-empty file is expected. `cargo` doesn't like empty manifests.

webern reviewed (commented):
Don't you need to remove `[tasks.check-rpm-structure]` (line 782 in the current version of your file)?

webern commented on twoliter/embedded/Makefile.toml:
```diff
@@ -829,6 +856,17 @@ export BUILDSYS_VARIANT_RUNTIME="${BUILDSYS_VARIANT_RUNTIME:?}"
+# compatible with previous versions.
+if [ -d "${BUILDSYS_ROOT_DIR}/variants/target/${BUILDSYS_ARCH}" ]; then
+  mkdir -p "${BUILDSYS_ROOT_DIR}/target"
+  mv \
```
Don't you also need to check if `"${BUILDSYS_ROOT_DIR}/Cargo.toml` exists?

bcressey replied:
> Don't you also need to check if "${BUILDSYS_ROOT_DIR}/Cargo.toml exists?

This works regardless of where the manifest workspace is located - it just relocates the targets output directory. That helps with downgrade compatibility since then `cargo` won't find it with an older `twoliter` release and will rebuild everything.

webern commented on twoliter/embedded/Makefile.toml:
```diff
@@ -1573,7 +1611,19 @@ rm -f ${BUILDSYS_TOOLS_DIR}/bin/*
-cargo clean --manifest-path "${BUILDSYS_ROOT_DIR}/Cargo.toml"
+# The workspace manifest could be the old one under variants, or the new one at
+# the project root.
+for ws in variants .; do
```
Oh, I see, I think I missed the `.` in a previous comment.

bcressey commented:
> Don't you need to remove [tasks.check-rpm-structure] (line 782 in the current version of your file)?

I can, but the `check-rpm-structure` task is never called, so we could also leave it alone for when we do want to make this a breaking change.

Commit:
twoliter: remove check-rpm-structure

This check is never called, and is no longer needed now that we can
work with either the flat or nested build/rpms directory layout.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

webern commented:
> > Don't you need to remove [tasks.check-rpm-structure] (line 782 in the current version of your file)?
> 
> I can, but the `check-rpm-structure` task is never called, so we could also leave it alone for when we do want to make this a breaking change.

Maybe we've headed that off by introducing a backwards compatible version and the `BUILDSYS_OUTPUT_GENERATION_ID` mechanism for a future break.

bcressey commented:
⬆️ switches from `ln *.rpm` to `find` to fix the case when one of the directories might be empty. `find` also lets us print out the matching RPMs, which is handy for troubleshooting.
